### PR TITLE
Build with and without icu variants

### PR DESCRIPTION
**Pull request by ocefpaf**
~~Just a test, for now...~~
See https://github.com/conda-forge/libxml2-feedstock/pull/121#issuecomment-2164644780

### CODE REVIEW

**Comment by mbargull** [comment]
`nm` diff
```diff
--- with_icu_yes
+++ with_icu_no
@@ -275,7 +274,0 @@
-U ucnv_close_73
-U ucnv_convertEx_73
-U UCNV_FROM_U_CALLBACK_STOP_73
-U ucnv_open_73
-U ucnv_setFromUCallBack_73
-U ucnv_setToUCallBack_73
-U UCNV_TO_U_CALLBACK_STOP_73
```
shows only the ICU imports, so this looks fine, as already predicted by @isuruf.
And taking a quick look/`grep`s at the source, only the signature of one struct is extended with ICU support at
https://gitlab.gnome.org/GNOME/libxml2/-/blob/v2.12.8/include/libxml/encoding.h#L138-L152
and that one is only ever passed on via its pointer type to ensure function declarations remain identical.

So, overall this looks good, technically.
Now it has to be decided if, how and when we'd incorporate this metadata-wise, i.e.,
- Is anyone opposed to this?
- What hacks for weighing down/choosing the variants do we want to offer? (`track_features`? build string prefixes?)
- Should we do this for `libxml2=2.12` or from  `libxml2>=2.13.0` onward? (The latter has been tagged just 17h ago :D.)

**Comment by ocefpaf** [comment]
> Is anyone opposed to this?

I suspect most people don't care. It was a niche addition back when I added it and I don't believe anyone else needed this. I may be wrong but, instead of waiting for comments that may never show up, I'm tempted to merge this and revert if it causes any breakage down the road.

> What hacks for weighing down/choosing the variants do we want to offer? (track_features? build string prefixes?)

No idea what is the best practice for this these days. Any suggestions?

> Should we do this for libxml2=2.12 or from libxml2>=2.13.0 onward? (The latter has been tagged just 17h ago :D.)

I prefer 2.13 so we can have a clean change. I'll update this PR to reflect that.

**Comment by ocefpaf** [comment]
> > Should we do this for libxml2=2.12 or from libxml2>=2.13.0 onward? (The latter has been tagged just 17h ago :D.)
> 
> I prefer 2.13 so we can have a clean change. I'll update this PR to reflect that.

OK. I'm not going to die on this hill. If anyone knows how to fix the Python build dependency here, please let us know. If not, I'll revert the commits and build the older version instead.

**Comment by ocefpaf** [comment]
@isuruf should we give up on 2.13.1 and build the current version instead?

**Comment by mbargull** [comment]
I'd really like to take the 2.13 bump to also purge the activation scripts.
Those shouldn't have been added it the first place for such a library.
We can probably go with functionality that upstream already provides (which would bake the prefix -- which wouldn't be too bad since it's already added to the binaries somewhere else; we can tidy it up later on).
Give me a few hours to get to take a look at this.

**Comment by ocefpaf** [comment]
> Give me a few hours to get to take a look at this.

This PR is 6 years in the making since #41. You have can few hours ;-p

**Comment by mbargull** [comment]
> This PR is 6 years in the making since #41. You have can few hours ;-p

Hehe, okay -- I also didn't really get to yesterday but will take a look today then.

**Comment by ocefpaf** [comment]
> Hehe, okay -- I also didn't really get to it yesterday but will take a look today then.

On a Saturday after midsommar? Isn't that illegal?

**Comment by mbargull** [comment]
> On a Saturday after midsommar? Isn't that illegal?

Haha, y'know, the days are starting to get shorter again -- I have to hustle to get a couple of things done ;P.
Hope you're enjoying the higher latitudes' summer (start and festivities)! I'm so happy it's summer again :).

**Comment by ocefpaf** [comment]
> Haha, y'know, the days are starting to get shorter again -- I have to hustle to get a couple of things done ;P.
> Hope you're enjoying the higher latitudes' summer (start and festivities)! I'm so happy it's summer again :).

Most people here wants to send me to jail when I say that I'm looking forward to winter. I'll probably regret it but I'm still, figuratively and literally, burnt by the SA sun.

Back to libxml2, the Windows failure is kind of hard to understand. I know that Windows has some odd return values, so maybe they are passing?

**Comment by ocefpaf** [comment]
@mbargull rebased, updated, and rerendered. Back to you ;-p

**Comment by gillins** [comment]
@ocefpaf @mbargull is this ready to go? Would be good to be able to merge this so we can update the versions.

**Comment by ocefpaf** [comment]
> @ocefpaf @mbargull is this ready to go? Would be good to be able to merge this so we can update the versions.

IMO, yes.

@mbargull anything you want to do before we merge this? From above:

```
I'd really like to take the 2.13 bump to also purge the activation scripts.
Those shouldn't have been added it the first place for such a library.
We can probably go with functionality that upstream already provides (which would bake the prefix -- which wouldn't be too bad since it's already added to the binaries somewhere else; we can tidy it up later on).
Give me a few hours to get to take a look at this.
```

Is that just removing the activation scripts or do we need to do more? I'm inclined to merge this tomorrow and leave this part for another iteration.

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -2,3 +2,7 @@ c_compiler:        # [osx]
 - clang_bootstrap  # [osx]
 cxx_compiler:      # [osx]
 - clang_bootstrap  # [osx]
+
+with_icu:
+- yes     # [not win]
+- no
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml:8]: We need to either keep windows compatible with ICU

File: recipe/meta.yaml [hunk 2]
```diff
@@ -32,17 +32,21 @@ requirements:
     - m2-patch  # [win]
     - libiconv  # [not win]
   host:
-    - icu  # [not win]
     - libiconv
+    - icu  # [with_icu == "yes"]
     - xz  # [not win]
     - zlib
+  run_constrained:
+    - icu <0.0a0  # [with_icu != "yes"]
```
Comment (h-vetinari) [inline recipe/meta.yaml:40]: ... or not apply the constraint on windows.

**Review by h-vetinari — COMMENTED** [review]
This causes problems on windows for any package depending on both libxml2 and icu:
```
The following packages are incompatible
\u251c\u2500 icu 75.1.* he0c23c2_0 is requested and can be installed;
\u2514\u2500 libpdal-core 2.8.1.* hd2cc1dd_1 is not installable because it requires
   \u2514\u2500 libxml2 >=2.13.3,<3.0a0 , which requires
      \u2514\u2500 icu <0.0a0 , which conflicts with any installable versions previously reported.
```

**Pull request merged**